source/Renderer/shaders/material_info.glsl

```diff
@@ -0,0 +1,304 @@
+
+// Get normal, tangent and bitangent vectors.
+NormalInfo getNormalInfo(vec3 v)
+{
+    vec2 UV = getNormalUV();
+    vec3 uv_dx = dFdx(vec3(UV, 0.0));
+    vec3 uv_dy = dFdy(vec3(UV, 0.0));
+
+    vec3 t_ = (uv_dy.t * dFdx(v_Position) - uv_dx.t * dFdy(v_Position)) /
+        (uv_dx.s * uv_dy.t - uv_dy.s * uv_dx.t);
+
+    vec3 n, t, b, ng;
+
+    // Compute geometrical TBN:
+#ifdef HAS_TANGENT_VEC4
+    // Trivial TBN computation, present as vertex attribute.
+    // Normalize eigenvectors as matrix is linearly interpolated.
+    t = normalize(v_TBN[0]);
+    b = normalize(v_TBN[1]);
+    ng = normalize(v_TBN[2]);
+#else
+    // Normals are either present as vertex attributes or approximated.
+#ifdef HAS_NORMAL_VEC3
+    ng = normalize(v_Normal);
+#else
+    ng = normalize(cross(dFdx(v_Position), dFdy(v_Position)));
+#endif
+    t = normalize(t_ - ng * dot(ng, t_));
+    b = cross(ng, t);
+#endif
+
+    // For a back-facing surface, the tangential basis vectors are negated.
+    if (gl_FrontFacing == false)
+    {
+        t *= -1.0;
+        b *= -1.0;
+        ng *= -1.0;
+    }
+
+    // Compute pertubed normals:
+#ifdef HAS_NORMAL_MAP
+    n = texture(u_NormalSampler, UV).rgb * 2.0 - vec3(1.0);
+    n *= vec3(u_NormalScale, u_NormalScale, 1.0);
+    n = mat3(t, b, ng) * normalize(n);
+#else
+    n = ng;
+#endif
+
+    NormalInfo info;
+    info.ng = ng;
+    info.t = t;
+    info.b = b;
+    info.n = n;
+    return info;
+}
+
+
+vec3 getClearcoatNormal(NormalInfo normalInfo)
+{
+#ifdef HAS_CLEARCOAT_NORMAL_MAP
+    vec3 n = texture(u_ClearcoatNormalSampler, getClearcoatNormalUV()).rgb * 2.0 - vec3(1.0);
+    n *= vec3(u_ClearcoatNormalScale, u_ClearcoatNormalScale, 1.0);
+    n = mat3(normalInfo.t, normalInfo.b, normalInfo.ng) * normalize(n);
+    return n;
+#else
+    return normalInfo.ng;
+#endif
+}
+
+
+vec4 getBaseColor()
+{
+    vec4 baseColor = vec4(1);
+
+#if defined(MATERIAL_SPECULARGLOSSINESS)
+    baseColor = u_DiffuseFactor;
+#elif defined(MATERIAL_METALLICROUGHNESS)
+    baseColor = u_BaseColorFactor;
+#endif
+
+#if defined(MATERIAL_SPECULARGLOSSINESS) && defined(HAS_DIFFUSE_MAP)
+    baseColor *= texture(u_DiffuseSampler, getDiffuseUV());
+#elif defined(MATERIAL_METALLICROUGHNESS) && defined(HAS_BASE_COLOR_MAP)
+    baseColor *= texture(u_BaseColorSampler, getBaseColorUV());
+#endif
+
+    return baseColor * getVertexColor();
+}
+
+
+void getSpecularGlossinessInfo(inout MaterialInfo info)
+{
+    info.f0 = u_SpecularFactor;
+    info.perceptualRoughness = u_GlossinessFactor;
+
+#ifdef HAS_SPECULAR_GLOSSINESS_MAP
+    vec4 sgSample = texture(u_SpecularGlossinessSampler, getSpecularGlossinessUV());
+    info.perceptualRoughness *= sgSample.a ; // glossiness to roughness
+    info.f0 *= sgSample.rgb; // specular
+#endif // ! HAS_SPECULAR_GLOSSINESS_MAP
+
+    info.perceptualRoughness = 1.0 - info.perceptualRoughness; // 1 - glossiness
+    info.c_diff = info.baseColor.rgb * (1.0 - max(max(info.f0.r, info.f0.g), info.f0.b));
+}
+
```

Comment:
IMO inout parameters should be avoided if possible as when reading the function call in the source code it's not explicitly clear that the function modifies the state of the input, and hence it's not explicitly clear what the function does

Reply:
has been addressed in b4411b4